### PR TITLE
CI and remove hard-coded product UID

### PR DESCRIPTION
**Pull request by m-mcgowan**

### CODE REVIEW

**Comment by m-mcgowan** [comment]
[Development Examples: remove hard-coded PRODUCT_UIDs from sample apps and example repos.](https://trello.com/c/gXVvrnf3/41-development-examples-remove-hard-coded-productuids-from-sample-apps-and-example-repos)

File: .github/workflows/note-arduino-ci.yml [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+    runs-on: ubuntu-latest
+    continue-on-error: true
+    strategy:
+      matrix:
+        example-sketch:
+          - ./energy/energy.ino
+        fully-qualified-board-name:
+          - STMicroelectronics:stm32:BluesW:pnum=SWAN_R5
+          # - esp32:esp32:featheresp32
+          # - adafruit:samd:adafruit_feather_m4
```
Comment (zfields) [inline .github/workflows/note-arduino-ci.yml]: Should these FQBNs be re-enabled?
```suggestion
          - esp32:esp32:featheresp32
          - adafruit:samd:adafruit_feather_m4
```

File: energy/energy.ino [hunk 2]
```diff
@@ -45,11 +49,13 @@ void setup()
...
-    J *req = notecard.newRequest("hub.set");
-    JAddStringToObject(req, "product", myProductID);
-    JAddStringToObject(req, "mode", "minimum");
-    notecard.sendRequest(req);
-
+    J *req;
+    if (myProductID[0]) {
+      req = notecard.newRequest("hub.set");
+      JAddStringToObject(req, "product", myProductID);
+      JAddStringToObject(req, "mode", "minimum");
```
Comment (zfields) [inline energy/energy.ino]: Looks like `"mode": "minimum"` should be offered to `hub.set`, regardless of `"product"` being set.

File: .github/actions/compile-examples/Dockerfile [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+ && arduino-cli core install adafruit:nrf52 \
+ && arduino-cli core install adafruit:samd \
+ && arduino-cli core install arduino:avr \
+ && arduino-cli core install arduino:mbed_nano \
+ && arduino-cli core install arduino:samd \
+ && arduino-cli core install esp32:esp32 \
+ && arduino-cli core install rp2040:rp2040 \
+ && arduino-cli core install SparkFun:apollo3 \
+ && arduino-cli core install STMicroelectronics:stm32 \
+ && arduino-cli lib install \"Adafruit BME680 Library\" \
```
Comment (zfields) [inline .github/actions/compile-examples/Dockerfile]: You are not using the BME680 library, and you can pull in the latest version of the Notecard library here instead.
```suggestion
 && arduino-cli lib install \"Blues Wireless Notecard\" \
```

File: .github/workflows/note-arduino-ci.yml [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+      - name: Checkout Code
+        id: checkout
+        uses: actions/checkout@v2
+        with:
+          submodules: true
+      - name: Compile Examples
+        id: compile_examples
+        uses: ./.github/actions/compile-examples
+        with:
+          libraries: 'Blues Wireless Notecard'
```
Comment (zfields) [inline .github/workflows/note-arduino-ci.yml]: Alternatively, this library could be baked into the Dockerfile as described above.

File: .github/actions/compile-examples/compile.sh [hunk 5]
```diff
@@ -0,0 +1,18 @@
+set -e
+SKETCH="$1"
+BOARD="$2"
+LIBS="$3"
+
+cp -r . /home/blues/Arduino/libraries/Blues_Wireless_Notecard
```
Comment (zfields) [inline .github/actions/compile-examples/compile.sh]: What is this step doing?

**Review by zfields — CHANGES_REQUESTED** [review]

File: energy/energy.ino (hunk 2 above)
Comment (m-mcgowan) [reply energy/energy.ino]: :+1: My first thought was that we'd leave it alone so not to surprise the developer, in the same way we don't want to mess with the `PRODUCT_UID`. But given this is an energy example, minimum is probably the better choice. I'll also update the readme so the developer is aware of the mode change.

File: .github/actions/compile-examples/compile.sh (hunk 5 above)
Comment (m-mcgowan) [reply .github/actions/compile-examples/compile.sh]: :+1: It was a hangover from the note-arduino CI

Comment (m-mcgowan) [reply .github/actions/compile-examples/compile.sh]: left over from copypasta from the note-arduino CI

File: .github/workflows/note-arduino-ci.yml (hunk 1 above)
Comment (m-mcgowan) [reply .github/workflows/note-arduino-ci.yml]: Thanks, I added these back in.

File: energy/energy.ino [hunk 6]
```diff
@@ -44,12 +49,13 @@ void setup()
...
 #endif
 
-    // Set the product UID and set the sync mode to be "minimum"
+    // Set the product UID if defined and set the sync mode to be "minimum"
     J *req = notecard.newRequest("hub.set");
-    JAddStringToObject(req, "product", myProductID);
-    JAddStringToObject(req, "mode", "minimum");
-    notecard.sendRequest(req);
-
+	if (myProductID[0]) {
```
Comment (zfields) [inline energy/energy.ino]: The code looks great, but this appears to be overly indented on GitHub.

Actually the whole file looks to be indented strangely.

Can you use VSCode to normalize it one way or the other?

**Review by zfields — APPROVED** [review]

**Pull request merged**